tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java

```diff
@@ -0,0 +1,192 @@
+        this.planRepository = planRepository;
+        this.usersRepository = usersRepository;
+        this.jsonTransformer = jsonTransformer;
+    }
+
+    @Override
+    public String getBasePath() {
+        return RATE_LIMIT_BASE_PATH;
+    }
+
+    @Override
+    public void define(Service service) {
+        service.post(ATTACH_PLAN_TO_USER_PATH, attachPlanToUser(), jsonTransformer);
+        service.get(GET_USERS_OF_PLAN_PATH, getUsersOfPlan(), jsonTransformer);
+        service.get(GET_PLAN_OF_USER_PATH, getPlanOfUser(), jsonTransformer);
+        service.delete(REVOKE_PLAN_OF_USER_PATH, revokePlanOfUser(), jsonTransformer);
+    }
+
+    private Route attachPlanToUser() {
+        return (request, response) -> {
+            Username username = extractUsername(request);
+            RateLimitingPlanId planId = extractPlanId(request);
+            userPreconditions(username);
+            planIdPreconditions(planId);
+            return Mono.from(planUserRepository.applyPlan(username, planId))
+                .then(Mono.just(Responses.returnNoContent(response)))
+                .onErrorResume(NullPointerException.class, e -> {
+                    throw ErrorResponder.builder()
+                        .statusCode(HttpStatus.BAD_REQUEST_400)
+                        .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
+                        .message(e.getMessage())
+                        .haltError();
+                })
+                .block();
+        };
+    }
+
+    private Route getUsersOfPlan() {
+        return (request, response) -> {
+            RateLimitingPlanId planId = extractPlanId(request);
+            planIdPreconditions(planId);
+            return Flux.from(planUserRepository.listUsers(planId))
+                .map(Username::asString)
+                .collectList()
+                .onErrorResume(NullPointerException.class, e -> {
+                    throw ErrorResponder.builder()
+                        .statusCode(HttpStatus.BAD_REQUEST_400)
+                        .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
+                        .message(e.getMessage())
+                        .haltError();
+                })
+                .block();
+        };
+    }
+
+    private Route getPlanOfUser() {
+        return (request, response) -> {
+            Username username = extractUsername(request);
+            userPreconditions(username);
+            return Mono.from(planUserRepository.getPlanByUser(username))
+                .map(planId -> new RateLimitingPlanIdResponse(planId.value().toString()))
+                .onErrorResume(NullPointerException.class, e -> {
+                    throw ErrorResponder.builder()
+                        .statusCode(HttpStatus.BAD_REQUEST_400)
+                        .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
+                        .message(e.getMessage())
+                        .haltError();
+                })
```

Comment:
I would enjoy that the code quality to be good enough for us not to need to catch NPEs at the presentation layer!